### PR TITLE
App Config Provider - Tag filters

### PR DESCRIPTION
**Pull request by mrm9084**
# Description

- Adds Tag Filters to provider
- SP fix in unit test

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [X] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 1]
```diff
@@ -48,11 +48,17 @@ class SettingSelector:
...
+    :type tag_filters: Optional[List[str]]
     """
 
-    def __init__(self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL):
+    def __init__(
+        self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL, tag_filters: Optional[List[str]] = None
+    ):
         self.key_filter = key_filter
         self.label_filter = label_filter
+        self.tag_filters = tag_filters
```
Comment (RichardChen820) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py:61]: Shall we check the format of each tag like what other languages are doing?
https://github.com/Azure/AppConfiguration-DotnetProvider/blob/8b883cce448004c901bb9ef3a855ee6576bd2281/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationOptions.cs#L231-L240

Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py:61]: Added

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_client_manager.py [hunk 2]
```diff
@@ -141,7 +141,7 @@ def load_configuration_settings(
         sentinel_keys = kwargs.pop("sentinel_keys", refresh_on)
         for select in selects:
             configurations = self._client.list_configuration_settings(
-                key_filter=select.key_filter, label_filter=select.label_filter, **kwargs
+                key_filter=select.key_filter, label_filter=select.label_filter, tags_filter=select.tag_filters, **kwargs
```
[the remark this answers is not in the document]
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_client_manager.py:144]: SDK uses `tags_filter`, but `tag_filters` is what we are using.

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 3]
```diff
@@ -48,11 +48,26 @@ class SettingSelector:
...
+
+        if tag_filters is not None:
```
Comment (albertofori) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py:61]: Should we add a check for maximum number of tags? There is a limit of 5

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 4]
```diff
@@ -48,11 +48,26 @@ class SettingSelector:
...
+    :type tag_filters: Optional[List[str]]
     """
 
-    def __init__(self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL):
+    def __init__(
+        self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL, tag_filters: Optional[List[str]] = None
+    ):
         self.key_filter = key_filter
         self.label_filter = label_filter
+        self.tag_filters = tag_filters
```
Comment (albertofori) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: nit: Should we validate before setting?

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 5]
```diff
@@ -56,18 +56,24 @@ class SettingSelector:
...
+            if len(tag_filters) > 5:
+                raise ValueError("tag_filters cannot be longer than 5 items.")
             for tag in tag_filters:
                 if not tag:
                     raise ValueError("Tag filter cannot be an empty string or None.")
                 if not isinstance(tag, str) or "=" not in tag or tag.startswith("="):
                     raise ValueError("Tag filter " + tag + ' does not follow the format "tagName=tagValue".')
+                forbidden_chars = {"*", "\\", ","}
+                if any(char in tag for char in forbidden_chars):
+                    raise ValueError("Tag filters cannot contain the '*', '\\', or ',' characters.")
```
Comment (albertofori) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: This check will flag [escaped](https://learn.microsoft.com/en-us/azure/azure-app-configuration/rest-api-key-value?pivots=v23-11#filtering:~:text=If%20a%20reserved%20character%20is%20part%20of%20the%20value%2C%20then%20it%20must%20be%20escaped%20by%20using%20%5C%7BReserved%20Character%7D.%20Non%2Dreserved%20characters%20can%20also%20be%20escaped.) reserved characters too, right?

Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: It should, I'm just checking if the character is in the string.

Comment (albertofori) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: Yeah, and that should not throw an error as there is a provision to accept escaped reserved characters if user wants to include them. I would say we either handle that scenario too or remove the reserved character validation and let the backend 400 bubble up

Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: Oh, I missed that part.

Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: It should be good now

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 6]
```diff
@@ -48,11 +48,70 @@ class SettingSelector:
...
+                if not isinstance(tag, str) or "=" not in tag or tag.startswith("="):
+                    raise ValueError("Tag filter " + tag + ' does not follow the format "tagName=tagValue".')
+                # Check for unescaped reserved characters
+                self._validate_tag_filter_escaping(tag)
+
         self.key_filter = key_filter
         self.label_filter = label_filter
+        self.tag_filters = tag_filters
+
+    def _validate_tag_filter_escaping(self, tag: str) -> None:
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: We dont need to validate this on client side. Server will validate and throw 400 if there are invalid characters

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 7]
```diff
@@ -48,11 +48,70 @@ class SettingSelector:
...
     """
 
-    def __init__(self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL):
+    def __init__(
+        self, *, key_filter: str, label_filter: Optional[str] = EMPTY_LABEL, tag_filters: Optional[List[str]] = None
+    ):
+        if tag_filters is not None:
+            if not isinstance(tag_filters, list):
+                raise TypeError("tag_filters must be a list of strings.")
+            if len(tag_filters) > 5:
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: This check not needed since we dont do the same check for comma separated keys either. Let server fail if more than 5 filters are given

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        """Test the validation of tag_filters in SettingSelector."""
+        # Valid tag filters
+        selector = SettingSelector(key_filter="*", tag_filters=["tag=value"])
+        assert selector.tag_filters == ["tag=value"]
+
+        selector = SettingSelector(key_filter="*", tag_filters=["tag=value", "another=one"])
+        assert selector.tag_filters == ["tag=value", "another=one"]
+
+        selector = SettingSelector(key_filter="*", tag_filters=["tag="])
+        assert selector.tag_filters == ["tag="]
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py:41]: Can you add test for tag with null value: "tag=\0"

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import pytest
+from azure.appconfiguration.provider._models import SettingSelector
+from azure.appconfiguration.provider._constants import EMPTY_LABEL
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py]: We can also offer EMPTY_TAG_VALUE in _constants file

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 10]
```diff
@@ -48,11 +48,70 @@ class SettingSelector:
     :keyword label_filter: A filter to select configuration settings based on their labels. Default is value is
      EMPTY_LABEL i.e. (No Label) as seen in the portal.
     :type label_filter: Optional[str]
+    :keyword tag_filters: A filter to select configuration settings based on their tags. This is a list of strings
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: Is this common between FF select options and kv select?

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py (hunk 6 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: @albertofori, thoughts? Your comment had me add this.

Comment (avanigupta) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: We skipped these validations in the .NET provider because the client cannot cover all the possible invalid character checks that server has. And we have precedent for not validating key/labels on the client side. We do a basic validation (like presence of `=` character and non-empty tag key). That's sufficient for client side.

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py (hunk 10 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: Yes. We do use the term configuration setting here and in the other ones. So it isn't super clear that it's for feature flags too.

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py (hunk 8 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py:41]: Is this meant to be a passing test? I can't get it to work, I get a bad request when creating it.

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py (hunk 10 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: @avanigupta do you have a suggestion to update the text?

Comment (avanigupta) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: How about - "A filter to select configuration settings and feature flags based on their tags."
This applies to all filter descriptions

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py (hunk 8 above)
Comment (avanigupta) [reply sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py:41]: Yes, this is a valid input. SDK should support `\0` as tag filter value and URL encode it as `%00` before making the request to AppConfig backend.

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py (hunk 9 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/tests/test_setting_selector.py]: Isn't empty tag just `None`? See my new change.

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py [hunk 11]
```diff
@@ -48,11 +48,27 @@ class SettingSelector:
...
+     that will be used to match tags on the configuration settings. Reserved characters (\\*, \\, ,) must be escaped
+     with backslash if they are part of the value.
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: It would be helpful to add an example of filtering tags with null value and empty value

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/testcase.py [hunk 12]
```diff
@@ -23,6 +23,7 @@ def create_aad_client(
         key_vault_options=None,
         on_refresh_success=None,
         feature_flag_enabled=False,
+        feature_flag_selectors=[SettingSelector(key_filter="*", label_filter="\0")],
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/tests/testcase.py]:
```suggestion
        feature_flag_selectors=[SettingSelector(key_filter="*", label_filter=EMPTY_LABEL)],
```

File: sdk/appconfiguration/azure-appconfiguration-provider/tests/testcase.py [hunk 13]
```diff
@@ -86,6 +90,7 @@ def create_client(
...
```
Comment (avanigupta) [inline sdk/appconfiguration/azure-appconfiguration-provider/tests/testcase.py]: Same here, if we have already defined EMPTY_LABEL for convenience, we should use it wherever "\0" is required.

File: sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py (hunk 11 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_models.py]: How is the new version.

**Review by avanigupta — APPROVED** [review]
